File: src-tauri/src/apis/jikan.rs
use serde::Deserialize;
use super::client::{ApiClient, ApiError};
use super::types::{ImageItem, CharacterItem};

const JIKAN_BASE_URL: &str = "https://api.jikan.moe/v4";

#[derive(Debug, Deserialize)]
struct JikanPicturesResponse {
    data: Vec<JikanPicture>,
}

#[derive(Debug, Deserialize)]
struct JikanPicture {
    jpg: JikanImageUrls,
    #[serde(default)]
    webp: Option<JikanImageUrls>,
}

#[derive(Debug, Deserialize)]
struct JikanImageUrls {
    image_url: Option<String>,
    small_image_url: Option<String>,
    large_image_url: Option<String>,
}

#[derive(Debug, Deserialize)]
struct JikanCharactersResponse {
    data: Vec<JikanCharacterData>,
}

#[derive(Debug, Deserialize)]
struct JikanCharacterData {
    character: JikanCharacter,
    role: String,
    #[serde(default)]
    voice_actors: Vec<JikanVoiceActor>,
}

#[derive(Debug, Deserialize)]
struct JikanCharacter {
    mal_id: i64,
    name: String,
    images: JikanCharacterImages,
}

#[derive(Debug, Deserialize)]
struct JikanCharacterImages {
    jpg: JikanImageUrls,
    #[serde(default)]
    webp: Option<JikanImageUrls>,
}

#[derive(Debug, Deserialize)]
struct JikanVoiceActor {
    person: JikanPerson,
    language: String,
}

#[derive(Debug, Deserialize)]
struct JikanPerson {
    mal_id: i64,
    name: String,
    images: JikanPersonImages,
}

#[derive(Debug, Deserialize)]
struct JikanPersonImages {
    jpg: JikanImageUrls,
}

pub struct JikanApi;

impl JikanApi {
    /// Get additional pictures for an anime by MAL ID
    pub async fn get_anime_pictures(
        client: &ApiClient,
        mal_id: i64,
    ) -> Result<Vec<ImageItem>, ApiError> {
        let url = format!("{}/anime/{}/pictures", JIKAN_BASE_URL, mal_id);

        let response = client.get("jikan", &url).await;

        let response = match response {
            Ok(r) => r,
            Err(ApiError::NotFound) => return Ok(vec![]),
            Err(e) => return Err(e),
        };

        let data: JikanPicturesResponse = response.json().await
            .map_err(|e| ApiError::InvalidResponse(e.to_string()))?;

        let images = data.data.into_iter()
            .filter_map(|pic| {
                let url = pic.jpg.large_image_url
                    .or(pic.jpg.image_url.clone())?;
                let thumb_url = pic.jpg.small_image_url
                    .or(pic.jpg.image_url)?;

                Some(ImageItem {
                    url,
                    thumb_url,
                    image_type: "picture".to_string(),
                    source: "jikan".to_string(),
                    width: None,
                    height: None,
                    vote_average: None,
                    likes: None,
                    language: None,
                })
            })
            .collect();

        Ok(images)
    }

    /// Get characters for an anime by MAL ID
    pub async fn get_anime_characters(
        client: &ApiClient,
        mal_id: i64,
    ) -> Result<Vec<CharacterItem>, ApiError> {
        let url = format!("{}/anime/{}/characters", JIKAN_BASE_URL, mal_id);

        let response = client.get("jikan", &url).await;

        let response = match response {
            Ok(r) => r,
            Err(ApiError::NotFound) => return Ok(vec![]),
            Err(e) => return Err(e),
        };

        let data: JikanCharactersResponse = response.json().await
            .map_err(|e| ApiError::InvalidResponse(e.to_string()))?;

        let characters = data.data.into_iter()
            .enumerate()
            .filter_map(|(i, char_data)| {
                let profile_url = char_data.character.images.jpg.image_url?;
                let profile_thumb_url = char_data.character.images.jpg.small_image_url
                    .unwrap_or_else(|| profile_url.clone());

                // Find Japanese voice actor
                let jp_va = char_data.voice_actors.iter()
                    .find(|va| va.language == "Japanese");

                Some(CharacterItem {
                    id: char_data.character.mal_id.to_string(),
                    name: char_data.character.name,
                    character_name: jp_va.map(|va| va.person.name.clone()),
                    profile_url,
                    profile_thumb_url,
                    order: i as i32,
                    source: "jikan".to_string(),
                    is_actor_headshot: false,
                })
            })
            .collect();

        Ok(characters)
    }

    /// Search for MAL ID from AniList ID (using title search as fallback)
    pub async fn find_mal_id(
        client: &ApiClient,
        title: &str,
    ) -> Result<Option<i64>, ApiError> {
        let url = format!(
            "{}/anime?q={}&limit=1",
            JIKAN_BASE_URL,
            urlencoding::encode(title)
        );

        let response = client.get("jikan", &url).await?;

        #[derive(Debug, Deserialize)]
        struct SearchResponse {
            data: Vec<SearchResult>,
        }
        #[derive(Debug, Deserialize)]
        struct SearchResult {
            mal_id: i64,
        }

        let data: SearchResponse = response.json().await
            .map_err(|e| ApiError::InvalidResponse(e.to_string()))?;

        Ok(data.data.first().map(|r| r.mal_id))
    }
}

File: scripts/content/core/api/jikan.ts
/**
 * Jikan API client (MyAnimeList unofficial API)
 * For additional anime pictures and characters
 * Uses REST - no API key required
 * Works in both Node.js and browser
 */

import type { ImageItem, CharacterItem } from '../types.js';

const JIKAN_BASE_URL = 'https://api.jikan.moe/v4';

// Rate limiting: Jikan has a 3 requests/second limit
let lastRequestTime = 0;
const MIN_REQUEST_INTERVAL = 350; // ms between requests

async function rateLimitedFetch(url: string): Promise<Response> {
	const now = Date.now();
	const timeSinceLastRequest = now - lastRequestTime;

	if (timeSinceLastRequest < MIN_REQUEST_INTERVAL) {
		await sleep(MIN_REQUEST_INTERVAL - timeSinceLastRequest);
	}

	lastRequestTime = Date.now();
	return fetch(url);
}

function sleep(ms: number): Promise<void> {
	return new Promise((resolve) => setTimeout(resolve, ms));
}

interface JikanPicturesResponse {
	data: {
		jpg: {
			image_url?: string;
			small_image_url?: string;
			large_image_url?: string;
		};
		webp?: {
			image_url?: string;
			small_image_url?: string;
			large_image_url?: string;
		};
	}[];
}

interface JikanCharactersResponse {
	data: {
		character: {
			mal_id: number;
			name: string;
			images: {
				jpg: {
					image_url?: string;
					small_image_url?: string;
				};
			};
		};
		role: string;
		voice_actors: {
			person: {
				mal_id: number;
				name: string;
				images: {
					jpg: {
						image_url?: string;
					};
				};
			};
			language: string;
		}[];
	}[];
}

interface JikanSearchResponse {
	data: {
		mal_id: number;
		title: string;
		images: {
			jpg: {
				image_url?: string;
				small_image_url?: string;
				large_image_url?: string;
			};
		};
		synopsis?: string;
		year?: number;
		season?: string;
		type?: string;
		status?: string;
		episodes?: number;
		score?: number;
		genres?: { name: string }[];
	}[];
}

/**
 * Get additional pictures for an anime by MAL ID
 */
export async function getAnimePictures(malId: number): Promise<ImageItem[]> {
	const url = `${JIKAN_BASE_URL}/anime/${malId}/pictures`;

	const response = await rateLimitedFetch(url);

	if (response.status === 404) {
		return [];
	}

	if (!response.ok) {
		throw new Error(`Jikan API error: ${response.status} ${response.statusText}`);
	}

	const data: JikanPicturesResponse = await response.json();

	return data.data
		.map((pic) => {
			const url = pic.jpg.large_image_url || pic.jpg.image_url;
			const thumbUrl = pic.jpg.small_image_url || pic.jpg.image_url;

			if (!url || !thumbUrl) return null;

			return {
				url,
				thumbUrl,
				imageType: 'picture',
				source: 'jikan'
			} as ImageItem;
		})
		.filter((img): img is ImageItem => img !== null);
}

/**
 * Get characters for an anime by MAL ID
 */
export async function getAnimeCharacters(malId: number): Promise<CharacterItem[]> {
	const url = `${JIKAN_BASE_URL}/anime/${malId}/characters`;

	const response = await rateLimitedFetch(url);

	if (response.status === 404) {
		return [];
	}

	if (!response.ok) {
		throw new Error(`Jikan API error: ${response.status} ${response.statusText}`);
	}

	const data: JikanCharactersResponse = await response.json();

	return data.data
		.map((charData, index) => {
			const profileUrl = charData.character.images.jpg.image_url;
			if (!profileUrl) return null;

			const profileThumbUrl = charData.character.images.jpg.small_image_url || profileUrl;

			// Find Japanese voice actor
			const jpVa = charData.voice_actors.find((va) => va.language === 'Japanese');

			return {
				id: charData.character.mal_id.toString(),
				name: charData.character.name,
				characterName: jpVa?.person.name,
				profileUrl,
				profileThumbUrl,
				order: index,
				source: 'jikan',
				isActorHeadshot: false
			} as CharacterItem;
		})
		.filter((char): char is CharacterItem => char !== null);
}

/**
 * Search for MAL ID by title
 */
export async function findMalId(title: string): Promise<number | null> {
	const url = `${JIKAN_BASE_URL}/anime?q=${encodeURIComponent(title)}&limit=1`;

	const response = await rateLimitedFetch(url);

	if (!response.ok) {
		throw new Error(`Jikan API error: ${response.status} ${response.statusText}`);
	}

	const data: JikanSearchResponse = await response.json();

	if (data.data.length === 0) {
		return null;
	}

	return data.data[0].mal_id;
}

/**
 * Search for anime by title (alternative to AniList)
 */
export async function searchAnime(
	query: string,
	limit: number = 15
): Promise<
	{
		malId: number;
		title: string;
		coverImage?: string;
		coverImageLarge?: string;
		synopsis?: string;
		year?: number;
		season?: string;
		type?: string;
		status?: string;
		episodes?: number;
		score?: number;
		genres: string[];
	}[]
> {
	const url = `${JIKAN_BASE_URL}/anime?q=${encodeURIComponent(query)}&limit=${limit}`;

	const response = await rateLimitedFetch(url);

	if (!response.ok) {
		throw new Error(`Jikan API error: ${response.status} ${response.statusText}`);
	}

	const data: JikanSearchResponse = await response.json();

	return data.data.map((item) => ({
		malId: item.mal_id,
		title: item.title,
		coverImage: item.images.jpg.small_image_url,
		coverImageLarge: item.images.jpg.large_image_url || item.images.jpg.image_url,
		synopsis: item.synopsis,
		year: item.year,
		season: item.season,
		type: item.type,
		status: item.status,
		episodes: item.episodes,
		score: item.score,
		genres: item.genres?.map((g) => g.name) || []
	}));
}
